### PR TITLE
[inductor] decompose broadcast-bias baddbmm into bmm + pointwise add

### LINKED ISSUE #187093: [inductor] baddbmm with broadcast bias: autotune never considers the bmm + pointwise-bias decomposition (1.3-1.45x left on the table at MoE expert-GEMM shapes)
### 🐛 Describe the bug

For `baddbmm` with a **broadcast** bias (`bias[B, 1, N]`, stride `[N, 0, 1]`
after broadcasting), Inductor's max-autotune chooses between (a) Triton
templates and (b) the extern ATen kernel — but never considers the
decomposition `bmm(x, w) + bias`, which is faster than both of its existing
choices whenever the bias add can fuse into a neighboring pointwise op
(which in MLP/MoE blocks it essentially always can — there is an activation
right after the GEMM).

Why both existing choices lose at this shape
(`[32, 640, 1024] x [32, 1024, 2048] + bias[32, 1, 2048]`, bf16, sm_103):

* **Extern `baddbmm` is honestly slow** (this is *not* an autotune
  measurement bug — we verified the benchmarked number is the true per-call
  cost): cuBLAS computes `C = alpha*A@B + beta*C`, so `at::baddbmm_out`
  must first materialize the broadcast bias into the output buffer. With a
  stride-0 bias dim that is a full `[B, M, N]`-sized strided elementwise
  copy (~75 us here), followed by a beta=1 GEMM that re-reads those bytes
  (~67 us): ~138 us per call, every call.
* **The Triton template wins the autotune for the op as written** (~104-106
  us; it reads the `[B, 1, N]` bias in its epilogue for free), so this is
  what Inductor ships.
* **The missing candidate is faster than both:** `bmm` routes extern to a
  beta=0 kernel (~60-62 us — C is never read or pre-filled), and the rank-1
  bias add fuses into the adjacent pointwise. Measured with the GELU
  consumer present (the realistic MLP/MoE shape of the graph):
  `gelu(baddbmm(...))` = 168.1 us/call vs `gelu(bmm(...) + bias)` =
  115.8 us/call — **1.45x** for the GEMM+activation pair. In the full MoE
  layer this came out to **1.31x end-to-end** (0.407 ms -> 0.311 ms median,
  verified numerics).

Kernel-level view (torch profiler, per call, GB300):

| graph as written | decomposed |
|---|---|
| `triton_tem_fused_baddbmm_0` 105.9 us | `nvjet_..._bz_NNT` (extern beta=0 bmm) 61.7 us |
| `triton_poi_fused_gelu_1` 69.3 us | `triton_poi_fused_add_gelu_0` 56.0 us |

Standalone (no consumer to fuse the bias add into), the decomposition is
*not* a win (102.2 vs 104.7 us here) — which is exactly why this wants to
be an autotune-level choice (benchmarked in graph context, where the
scheduler knows whether the epilogue can fuse) or a decomposition with a
consumer-aware condition, rather than an unconditional rewrite.

Possibly related: the ATen-level cost itself (broadcast-bias `baddbmm`
paying a full output-sized materialization copy: 138.0 us vs 60.3 us for
the bare `bmm`, eager) could separately be improved in cuBLAS dispatch
(e.g. via a fused-epilogue path), but the Inductor decomposition choice is
the more general fix and is what this issue asks for.

### Repro

Standalone script attached (`repro.py`, no external deps). It compiles both
formulations with `mode="max-autotune-no-cudagraphs"`, prints per-call
timings + the kernels each compiled graph runs, and asserts the template
choice and the gap. Output on a GB300 (sm_103):

```
eager baddbmm (broadcast bias [B,1,N]):    138.0 us/call
eager bmm     (no bias)               :     60.3 us/call
compiled baddbmm        (bare)        :    102.2 us/call
compiled bmm + bias     (bare)        :    104.7 us/call
compiled gelu(baddbmm)  (as written)  :    168.1 us/call
compiled gelu(bmm+bias) (decomposed)  :    115.8 us/call
decomposition speedup with consumer   :    1.451x
```

### Suggested fix

Register a decomposition (or an additional autotune candidate) for
`baddbmm` when the bias input has a stride-0 broadcast dim:
`bmm(batch1, batch2) + input`. The pointwise add participates in normal
Inductor fusion, so wherever the GEMM output feeds a pointwise op the bias
is free; the autotuner's existing benchmarking would keep the current
behavior in the (rarer) standalone case if the candidate loses there.

### Versions

* torch 2.12.0a0 (NGC 26.05 container build), CUDA 13.2, driver 580.159.03
* GPU: NVIDIA GB300 (sm_103); shapes are MoE expert-GEMM class
  (`E=32` experts, `tokens*capacity=640`, `hidden=1024`, `ffn=2048`, bf16)
* Reproduces with caches disabled (`TORCHINDUCTOR_FORCE_DISABLE_CACHES=1`)

<details>
<summary><code>repro.py</code> (standalone, no external deps)</summary>

```python
# Standalone reproducer: torch.compile (Inductor, max-autotune GEMM tuning)
# never considers the `bmm + pointwise bias` decomposition as an autotune
# choice for `baddbmm` with a BROADCAST bias, leaving measurable performance
# on the table at MoE expert-GEMM shapes.
#
# Mechanism (all autotune numbers honest, none of this is a benchmark bug):
#   * For `baddbmm(bias[B,1,N], x[B,M,K], w[B,K,N])` the extern (cuBLAS/ATen)
#     choice is honestly slow: cuBLAS computes C = alpha*A@B + beta*C, so
#     `at::baddbmm_out` must first MATERIALIZE the broadcast bias into the
#     output buffer (a strided elementwise copy over the full [B,M,N] output)
#     and then run a beta=1 GEMM that re-reads it.
#   * The Triton template, which reads the [B,1,N] bias in its epilogue for
#     free, therefore wins the autotune FOR THE OP AS WRITTEN.
#   * But the decomposition `bmm(x, w) + bias` is faster than BOTH: the pure
#     GEMM routes extern to a beta=0 kernel (C never read), and the rank-1
#     bias add is a cheap pointwise kernel that Inductor can fuse into
#     neighboring ops in real graphs (where it becomes effectively free).
#
# This script compiles both forms at a representative shape, both bare and
# with the activation that follows the GEMM in any MLP/MoE block (GELU),
# prints the kernels each compiled graph actually runs, and asserts:
#   1. compiled baddbmm keeps a Triton GEMM template; the decomposed form
#      routes the GEMM to an extern beta=0 kernel (no Triton template),
#   2. with the activation consumer present (the realistic case), the
#      decomposed form is faster end-to-end: the bias add fuses into the
#      activation pointwise for free,
#   3. eager baddbmm with a broadcast bias is much slower than eager bmm
#      (the ATen materialization cost itself).
#
# The bare (no-consumer) pair is also timed and printed: standalone, the
# unfused bias add costs more than the template's free epilogue read, which
# is exactly why this should be an AUTOTUNE-LEVEL DECOMPOSITION CHOICE
# (benchmarked in graph context, where pointwise fusion applies) rather
# than an unconditional decomposition.
#
# Tested environment: NVIDIA GB300 (sm_103), CUDA 13.2, driver 580.159.03,
# torch 2.12.0a0 (NGC 26.05). No repo dependencies.

import os

# Fresh tune every run: keep the experiment hermetic.
os.environ.setdefault("TORCHINDUCTOR_FORCE_DISABLE_CACHES", "1")

import torch  # noqa: E402

B, M, K, N = 32, 640, 1024, 2048  # MoE expert GEMM1 class (bf16)
DTYPE = torch.bfloat16
WARMUP, ITERS = 20, 200


def cuda_time_us(fn, *args):
    for _ in range(WARMUP):
        fn(*args)
    torch.cuda.synchronize()
    beg = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    beg.record()
    for _ in range(ITERS):
        fn(*args)
    end.record()
    torch.cuda.synchronize()
    return beg.elapsed_time(end) * 1000.0 / ITERS


def gpu_kernels(fn, *args):
    """Names + total us of CUDA kernels per call (5-call profile)."""
    from torch.profiler import ProfilerActivity, profile

    fn(*args)  # ensure compiled
    torch.cuda.synchronize()
    with profile(activities=[ProfilerActivity.CUDA]) as prof:
        for _ in range(5):
            fn(*args)
        torch.cuda.synchronize()
    out = {}
    for evt in prof.key_averages():
        if evt.device_type == torch.autograd.DeviceType.CUDA and evt.self_device_time_total > 0:
            out[evt.key] = out.get(evt.key, 0.0) + evt.self_device_time_total / 5.0
    return out


def op_baddbmm(bias, x, w):
    return torch.baddbmm(bias, x, w)


def op_bmm_bias(bias, x, w):
    return torch.bmm(x, w) + bias


def op_baddbmm_gelu(bias, x, w):
    return torch.nn.functional.gelu(torch.baddbmm(bias, x, w))


def op_bmm_bias_gelu(bias, x, w):
    return torch.nn.functional.gelu(torch.bmm(x, w) + bias)


def main():
    assert torch.cuda.is_available()
    torch.manual_seed(42)
    dev = "cuda"
    x = torch.randn(B, M, K, device=dev, dtype=DTYPE)
    w = torch.randn(B, K, N, device=dev, dtype=DTYPE)
    bias = torch.randn(B, 1, N, device=dev, dtype=DTYPE)  # broadcast over M

    # --- eager: the ATen-level cost of the broadcast-bias materialization ---
    t_eager_baddbmm = cuda_time_us(op_baddbmm, bias, x, w)
    t_eager_bmm = cuda_time_us(torch.bmm, x, w)
    print(f"eager baddbmm (broadcast bias [B,1,N]): {t_eager_baddbmm:8.1f} us/call")
    print(f"eager bmm     (no bias)               : {t_eager_bmm:8.1f} us/call")

    # --- compiled: what Inductor picks for each formulation ---
    compiled_baddbmm = torch.compile(
        op_baddbmm, fullgraph=True, dynamic=False, mode="max-autotune-no-cudagraphs"
    )
    compiled_bmm_bias = torch.compile(
        op_bmm_bias, fullgraph=True, dynamic=False, mode="max-autotune-no-cudagraphs"
    )

    t_c_baddbmm = cuda_time_us(compiled_baddbmm, bias, x, w)
    t_c_bmm_bias = cuda_time_us(compiled_bmm_bias, bias, x, w)
    print(f"compiled baddbmm        (bare)        : {t_c_baddbmm:8.1f} us/call")
    print(f"compiled bmm + bias     (bare)        : {t_c_bmm_bias:8.1f} us/call")
    print("  (bare, the unfused bias add costs more than the template's free"
          " epilogue read — the decomposition only pays in graph context)")

    # --- compiled, with the activation consumer (the realistic MLP/MoE case)
    compiled_baddbmm_gelu = torch.compile(
        op_baddbmm_gelu, fullgraph=True, dynamic=False,
        mode="max-autotune-no-cudagraphs",
    )
    compiled_decomposed_gelu = torch.compile(
        op_bmm_bias_gelu, fullgraph=True, dynamic=False,
        mode="max-autotune-no-cudagraphs",
    )
    t_c_baddbmm_gelu = cuda_time_us(compiled_baddbmm_gelu, bias, x, w)
    t_c_decomposed_gelu = cuda_time_us(compiled_decomposed_gelu, bias, x, w)
    print(f"compiled gelu(baddbmm)  (as written)  : {t_c_baddbmm_gelu:8.1f} us/call")
    print(f"compiled gelu(bmm+bias) (decomposed)  : {t_c_decomposed_gelu:8.1f} us/call")
    print(f"decomposition speedup with consumer   : "
          f"{t_c_baddbmm_gelu / t_c_decomposed_gelu:8.3f}x")

    # numerics: identical inputs, compare outputs
    ref = op_baddbmm_gelu(bias.float(), x.float(), w.float())
    out_a = compiled_baddbmm_gelu(bias, x, w).float()
    out_b = compiled_decomposed_gelu(bias, x, w).float()
    for name, out in (("gelu(baddbmm)", out_a), ("gelu(bmm+bias)", out_b)):
        rel = (out - ref).abs().max() / ref.abs().max()
        print(f"max rel err vs fp32 ref, {name:15s}: {rel.item():.3e}")
        assert rel < 2e-2, f"{name} numerics off"

    kerns_a = gpu_kernels(compiled_baddbmm_gelu, bias, x, w)
    kerns_b = gpu_kernels(compiled_decomposed_gelu, bias, x, w)
    print("\nkernels, compiled gelu(baddbmm):")
    for k, v in sorted(kerns_a.items(), key=lambda i: -i[1]):
        print(f"  {v:9.1f} us  {k}")
    print("kernels, compiled gelu(bmm + bias):")
    for k, v in sorted(kerns_b.items(), key=lambda i: -i[1]):
        print(f"  {v:9.1f} us  {k}")

    # 1. template choice: baddbmm keeps a Triton template GEMM; the
    #    decomposition routes the GEMM extern (cuBLAS), no template survives.
    a_has_template = any("triton_tem" in k for k in kerns_a)
    b_has_template = any("triton_tem" in k for k in kerns_b)
    assert a_has_template, (
        "expected compiled gelu(baddbmm) to pick a Triton GEMM template "
        f"(got: {sorted(kerns_a)})"
    )
    assert not b_has_template, (
        "expected compiled gelu(bmm+bias) to route the GEMM extern "
        f"(got: {sorted(kerns_b)})"
    )

    # 2. with the consumer present, the decomposition is faster end-to-end.
    assert t_c_baddbmm_gelu / t_c_decomposed_gelu > 1.05, (
        f"decomposition not faster with consumer: "
        f"{t_c_baddbmm_gelu:.1f} vs {t_c_decomposed_gelu:.1f} us"
    )

    # 3. ATen-level: broadcast-bias baddbmm pays a full output-sized
    #    materialization copy before the GEMM.
    assert t_eager_baddbmm / t_eager_bmm > 1.5, (
        f"expected eager broadcast-bias baddbmm >> eager bmm: "
        f"{t_eager_baddbmm:.1f} vs {t_eager_bmm:.1f} us"
    )

    print("\nALL ASSERTIONS PASSED — Inductor lacks the bmm+pointwise-bias "
          "decomposition choice for broadcast-bias baddbmm.")


if __name__ == "__main__":
    main()

```
</details>

*Suggested labels: `module: inductor`, `topic: performance`*

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by cfregly**
## Issue

Fixes #187093

## Summary

For `baddbmm` with a bias broadcast along dim 1 (M) of the `[B, M, N]` output, register a conditional Inductor decomposition to `bmm(batch1, batch2) + bias`, following the existing conditional `addmm`/`mm`/`bmm` decomps in `torch/_inductor/decomposition.py` (returns `NotImplemented` when the condition does not hold). Mechanism and design discussion are in the issue: the extern choice pays a full output-sized bias materialization copy plus a beta=1 GEMM re-read; the Triton template that wins the autotune as-written is still slower than extern beta=0 `bmm` + a pointwise bias add that fuses into the neighboring pointwise op (the activation, in MLP/MoE blocks).

Measured on GB300 (sm_103), bf16, `[32, 640, 1024] x [32, 1024, 2048] + bias[32, 1, 2048]`, `max-autotune-no-cudagraphs` (per-call, torch profiler):

| | graph as written (before) | decomposed (after) |
|---|---|---|
| GEMM | `triton_tem_fused_baddbmm_0` 105.9 us | `nvjet_...bz_NNT` (extern beta=0 bmm) 61.7 us |
| pointwise | `triton_poi_fused_gelu_1` 69.3 us | `triton_poi_fused_add_gelu_0` 56.0 us |
| `gelu(baddbmm)` total | 168.1 us | 115.8 us (**1.45x**) |

1.31x end-to-end at the MoE layer this came from (0.407 ms -> 0.311 ms median, verified numerics). Re-validated with this exact patch applied to a GB300 installed torch (2.12.0a0): the as-written `gelu(baddbmm(...))` now fires `decompose_baddbmm`, compiles to the extern beta=0 `nvjet` bmm + a single fused bias-add+gelu pointwise kernel, and matches the manually decomposed formulation's per-call time; max rel err vs fp32 eager ref 5.6e-03 (same as before the patch).

Standalone (no pointwise consumer to fuse into) the two forms are within a few percent (102 vs 105 us at this shape), which is why the decomposition is gated on the broadcast-M condition only; a full `[B, M, N]` bias keeps the tuned baddbmm lowering unchanged (verified: counter stays 0, extern/template choice as before). Semantics: beta=0 ignores the bias entirely (no NaN propagation, matching eager and the existing `test_baddbmm`), beta/alpha scaling preserved, `out_dtype`/cpu/mps/integer inputs fall through to the existing lowering, and unlike the excluded core-ATen baddbmm decomposition the GEMM stays in the input dtype (no opmath upcast).

Validation run on GB300 against the patched installed torch: `test/inductor/test_torchinductor.py -k baddbmm` (3 passed, 1 skipped), `test_torchinductor_dynamic_shapes.py -k baddbmm` (3 passed, 1 skipped), `test_torchinductor_codegen_dynamic_shapes.py -k baddbmm` (1 passed, 2 xfailed as before), `test_max_autotune.py -k baddmm` (4 passed, including the broadcast-bias `triton_tem_fused_baddbmm` FileCheck, whose kernel name is preserved through the decomposition via `original_aten`), and a baddbmm OpInfo sample sweep across fp32/fp16/bf16 (18/18 pass, 12 taking the new path).

## Checklist

- [x] Passes lint (`lintrunner` on the changed files; PYREFLY skipped — not runnable in an unbuilt checkout)
- [x] Added/updated tests (`test_baddbmm_broadcast_bias_decompose`: decomposition fires for broadcast bias, not for full bias, beta=0 NaN semantics, beta/alpha scaling)
- [ ] Updated documentation (not applicable)
- [x] Included benchmark results (table above, from the issue + re-validated with this patch)

## BC-breaking?

No.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related to this PR, and review the changes

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Review by eellison — COMMENTED** [review]
This is very similar to https://github.com/pytorch/pytorch/blob/e1e9750ea1041259c0b691bd59d4ac10e19ae82c/torch/_inductor/fx_passes/post_grad.py#L1734-L1762. 

Should we just extend that pass there ? note, this changes numerics, so it's off by default for dtypes < fp32
